### ISSUE TITLE
Add a textbox for numeric values.

### ISSUE THREAD

**Issue by JAicewizard** [linebender/druid#1140]
Right now there is no way to have custom verify functions, or prevent people from typing in non-numeric characters.
It would be great if there was some kind of way to have textboxes that only allow numeric content.
It would be amazing if this could be implemented in such a way where the textbox implements `Widget<uxx/fxx>`

**Comment by Finnerale** [comment]
Maybe [`Parse`](https://docs.rs/druid/0.6.0/druid/widget/struct.Parse.html) is what you are looking for? There also is an [example](https://github.com/linebender/druid/blob/master/druid/examples/parse.rs) for it.

**Comment by JAicewizard** [comment]
It is sort of, but it doesn't prevent anyone from typing in non-numeric characters.
I can still type in anything, it will just result in `None(0)` as data.

**Comment by Finnerale** [comment]
Hmm, in that case maybe a `Lens` filtering out non-numeric characters could work?
This is what I would imagine (just a sketch, haven't tested it):
```rust
struct NumericLens;
impl Lens<String, String> for NumericLens {
    fn with<V, F: FnOnce(&String) -> V>(&self, data: &String, f: F) -> V {
        f(data)
    }
    fn with_mut<V, F: FnOnce(&mut String) -> V>(&self, data: &mut String, f: F) -> V {
        let v = f(data);
        data.retain(|c| c.is_digit(10));
        v
    }
}
```

**Comment by JAicewizard** [comment]
I got something to work using:
```
struct StringLens;
impl Lens<AppState, String> for StringLens {
    fn with<V, F: FnOnce(&String) -> V>(&self, data: &AppState, f: F) -> V {
        let mut string = data.icons.to_string();
        if data.icons == 0 {
            string = "".into();
        }
        f(&string)
    }
    fn with_mut<V, F: FnOnce(&mut String) -> V>(&self, data: &mut AppState, f: F) -> V {
        let mut string = data.icons.to_string();
        if data.icons == 0 {
            string = "".into();
        }
        let v = f(&mut string);
        string.retain(|c| c.is_digit(10));
        data.icons = string.parse().unwrap_or(0);
        v
    }
}
```

This does exactly what I want, it also allows me to have a slider point to the same value.
It is however not very elegant, and numeric input is something a lot of applications will end up needing in some way or another.
I would leave this issue open for future development, but if you feel this is good enough support feel free to close this.
If you do close this, it would be nice to have an example specifically for this, possibly linked in the textbox documentation.

**Comment by cmyr** [comment]
A data validating textbox is definitely something we want eventually; I have a version of this (as `EditableLabel`) in runebender.

**Comment by MrGibus** [comment]
I'd imagine this can already be done using a controller with some fiddling?

**Comment by cmyr** [comment]
Fixed with #1377
